druid-shell/src/platform/mac/application.rs

```diff
@@ -32,57 +34,91 @@ use crate::application::AppHandler;
 
 static APP_HANDLER_IVAR: &str = "druidAppHandler";
 
-pub struct Application {
+#[derive(Clone)]
+pub(crate) struct Application {
     ns_app: id,
+    state: Rc<RefCell<State>>,
+}
+
+struct State {
+    quitting: bool,
 }
 
 impl Application {
-    pub fn new(handler: Option<Box<dyn AppHandler>>) -> Application {
+    pub fn new() -> Application {
+        // macOS demands that we run not just on one thread,
+        // but specifically the first thread of the app.
         util::assert_main_thread();
         unsafe {
             let _pool = NSAutoreleasePool::new(nil);
 
-            let delegate: id = msg_send![APP_DELEGATE.0, alloc];
```

Comment:
this stuff was very carefully organized after much trial and error, and based on a close reading of how appkit starts an application; I would leave it as is unless there's a clear problem.


edit: I think I see your rationale; you would like `run` to be symmetric. Do we imagine a world where `run` can be called multiple times? Wouldn't you need to call `new` between calls to `quit` and `run` in this case, anyway? I think this probably should be in the `run` docs, in any case.

Reply:
The delegate was previously set after creating the app but before running it - and that is still the case now. I just moved it from `new` to `run`. There's no actual change to the order of events though. On my macOS testing it seems to work fine, including the `applicationDidFinishLaunching` event.

I can imagine `run` being called twice by accident, but that will panic on purpose. The check is done in the platform-independent `Application::run` which also documents this. Also yes, you can't call `run` on the same exact copy of `Application` because `run` consumes `self`. That's one safe-guard. However because `Application` is clonable someone might for some bizarre reason call `run` on that clone elsewhere. That's why there's a check for it which will panic.